### ISSUE TITLE
Obtain feature-level attention instead of grouped attention?

### ISSUE THREAD

**Issue by mizzlefeng** [PriorLabs/TabPFN#885]
Hello,

I am currently using **TabPFN v2.1.3** and I would like to ask about how to obtain **feature-level attention weights** from the model.

I noticed the reply in issue #258:

> “Looking at the cosine similarity between embedding will give you a related metric but not exactly the attention weights.
> To get the attention weights, you can look at the `.model_` attribute (created after fit) and add PyTorch hooks to it (you can see the model architecture in `tabpfn.model.transformer`).”

This helped me identify where the attention modules are located in the source code. However, when I tried to register hooks and inspect the model, I found that:

```python
clf.model_.features_per_group
```

returns **2**.

This seems consistent with the statement in the paper that **“a transformer position encodes two features of one example.”**

Because of this, I am confused about the granularity of the extracted attention:

* Does this mean that, even if I hook the attention modules successfully, I can only obtain **attention over grouped feature sets**, rather than **attention for each original feature**?
* In other words, are the attention weights fundamentally defined at the **feature-group/token level** instead of the **raw feature level**?
* If so, is there any official or recommended way to map these grouped attention weights back to the original features?
* Or is feature-level attention simply not directly recoverable from the current architecture?

My goal is to interpret the model at the **individual feature level**, rather than only at the grouped-token level.

Any clarification would be greatly appreciated. Thank you very much.

**Comment by adrian-prior** [comment]
Hi @mizzlefeng, your reading is correct.

  Features are grouped *before* attention in [`transformer.py` (L389–L395)](https://github.com/PriorLabs/TabPFN/blob/fded30e/src/tabpfn/architectures/base/transformer.py#L389-L395):

  ```python
  # Splits up the input into subgroups
  x[k] = einops.rearrange(
      x[k],
      "s b (f n) -> b s f n",
      n=self.features_per_group,
  )
  ```

  Every token the transformer sees is a **contiguous group of `features_per_group` raw features** projected through a small encoder. Attention weights are defined over these tokens, not over individual features.

  The grouping size is baked into each checkpoint's config, and there is no way to directly map the attention weights back to the feature level.

  A handful of older v2 ensemble-member checkpoints were trained with `features_per_group=1`, which makes each token correspond to exactly one raw feature. These are published on HuggingFace alongside the default checkpoint. Example:

  ```python
  from huggingface_hub import hf_hub_download
  from tabpfn import TabPFNClassifier

  # Older v2 ensemble-member checkpoint trained with features_per_group=1
  ckpt_path = hf_hub_download(
      repo_id="Prior-Labs/TabPFN-v2-clf",
      filename="tabpfn-v2-classifier-gn2p4bpt.ckpt",
  )

  clf = TabPFNClassifier(model_path=ckpt_path, n_estimators=1)
  clf.fit(X_train, y_train)

  print(clf.model_.features_per_group)  # -> 1
  ```

Note that these older checkpoints are superseded by newer model versions and the default v2 checkpoints, though.

If you are looking into this from an interpretability perspective, I can also recommend looking into the interpretability tools we provide in `tabpfn-extensions`: 

  - [Interpretability extension docs & code](https://github.com/PriorLabs/tabpfn-extensions/tree/main/src/tabpfn_extensions/interpretability)
  - [Example — SHAP values on TabPFN](https://github.com/PriorLabs/tabpfn-extensions/blob/main/examples/interpretability/shap_example.py)

Hope this clarifies things!
